File: y15/d10/cpp/main.cpp
#include "helper.h"
#include <iostream>
#include <ostream>
#include <string>
#include <vector>
using namespace std;

string look_and_say(const string &sequence) {
  char current_digit = sequence[0];
  int count = 0;
  string result = "";

  for(const auto& d: sequence) {
    if (current_digit == d) {
      count++;
    } else {
      result += to_string(count);
      result.push_back(current_digit);

      current_digit = d;
      count = 1;
    }
  }
  //add the last one
  result += to_string(count);
  result.push_back(current_digit);

  return result;
}

string evolve_sequence(string sequence, int generation) {
  for (size_t i = 0; i < generation; i++) {
    sequence = look_and_say(sequence);
  }
  return sequence;
}

int main(int argc, char *argv[]) {
  if (argc < 2) {
    cout << "Provide the input file name" << endl;
    return 1;
  }

  vector<string> lines = read_lines(argv[1]);
  string sequence = lines[0];

  string part1 = evolve_sequence(sequence, 40);
  cout << "Part 1: " << part1.size() << endl;

  string part2 = evolve_sequence(sequence, 50);
  cout << "Part 2: " << part2.size() << endl;

  return 0;
}

File: y15/d10/golang/main.go
package main

import (
	"flag"
	"fmt"
	"strconv"
)

func main() {
	f := flag.String("f", "", "Input File Name")
	flag.Parse()
	lines := readLines(*f)
	solve(lines)
}

// Why bytes? String is immutable, tried and failed.
// slices were easier to handle
func lookAndSay(sequence []byte) []byte {
	result := make([]byte, 0)
	currentDigit := sequence[0]
	count := 0
	for _, d := range sequence {
		if currentDigit == d {
			count++
		} else {
			result = append(result, strconv.Itoa(count)...)
			result = append(result, currentDigit)

			currentDigit = d
			count = 1
		}
	}
	result = append(result, strconv.Itoa(count)...)
	result = append(result, currentDigit)

	return result
}

func evolveSequence(sequence []byte, generation int) []byte {
	for range generation {
		sequence = lookAndSay(sequence)
	}
	return sequence
}

func solve(lines []string) {
	sequence := []byte(lines[0])

	part1 := evolveSequence(sequence, 40)
	fmt.Println("Part 1:", len(part1))

	part2 := evolveSequence(sequence, 50)
	fmt.Println("Part 2:", len(part2))
}
